### ISSUE TITLE
Support for phi3 small and medium?

### ISSUE THREAD

**Issue by Optimox** [pytorch/torchtune#1013]
Microsoft released weights for two larger versions of phi3 with phi3-small (7B) and phi3-medium (14B).

Is there someone already working on this?

**Comment by kartikayk** [comment]
Thanks for opening this issue!

We don't have anyone working on these at the moment. Would you be interested in adding support?

**Comment by Optimox** [comment]
I'll try to find some time to have a look, I'll let you know!

**Comment by Optimox** [comment]
Should torchtune be independent of triton ?

**Comment by ebsmothers** [comment]
So far we have not taken any explicit triton dependency in our library. Is there a kernel you would need to add as part of the phi3 work?

**Comment by Optimox** [comment]
It looks like phi3 medium would be very easy to add as the architecture is the same as mini just with different parameters. However, the phi3 small architecture is different and seems to rely on triton: https://huggingface.co/microsoft/Phi-3-small-8k-instruct/blob/main/triton_blocksparse_attention_layer.py

**Comment by ebsmothers** [comment]
@Optimox thanks for looking into this. In addition to the triton dependency I think there is also a flash-attention dependency in the HF version of the repo, which is probably something we don't want to take on at this time. I haven't looked closely enough yet, but I'm wondering if it's possible to have an equivalent version without these dependencies.. I know @RdoubleA had looked at taking a flash-attention dependency as part of his work on sample packing but ultimately decided against it. 

(To clarify we already do take a dependency on triton by virtue of installing pytorch, but I think it's a separate question whether we want to explicitly use Triton kernels in our modeling code. This is something we haven't done yet and I'd be reluctant to do it as a one-off here without thinking through the implications a bit more.)

Would you be interested in adding just the medium model for the short-term, and in parallel you/we can look into the best path for supporting the small model? (Admittedly I need to do some more homework here myself.)

**Comment by Optimox** [comment]
@ebsmothers I personnally agree, I love the fact that torchtune is in pure pytorch, it starts to be too complicated if it uses very recent external libraries. I'll finish the gemma7b PR and then start the phi3 small support.

**Comment by ebsmothers** [comment]
Belatedly coming back to this. Now that Phi4 is out and we have a PR to support it (#2197), I think this is no longer relevant
